### PR TITLE
feature/index management

### PR DESCRIPTION
**Pull request by inzanez**
Initial implementation suggestion for index management on ArangoDB.

### CODE REVIEW

File: src/index.rs [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DeleteIndexResponse {
+    pub id: String,
+    pub error: bool,
+    pub code: u16,
+}
+
+impl Index {
+    fn default(r#type: Type) -> Self {
```
Comment (fMeow) [inline src/index.rs:83]: I prefer implementing `Default` trait in this case.

**Comment by fMeow** [comment]
1. I prefer use varaible name for getter without a `get` prefix. `get_index ` and `get_indexes` can simply be `index` and `indexes`.
2. We don't need `error` and `code` field in response type. The helper function `deserialize_response` has already map the response to `ArangoError` if the `error` field is true.
3. It seems that some index type has their own fields. For instance, a geo index only has `fields` while ttl index has an additional `expire_after` field. Can we use a fat enum to ensure that a geo index must not has a `expire_after` field and etc?

File: src/index.rs (hunk 1 above)
Comment (EpiFouloux) [reply src/index.rs:83]: I think it's because he can't randomy pick the index type, the default is relative to the index type here (`r#type` argument)

**Comment by EpiFouloux** [comment]
thank you very much for this

**Comment by inzanez** [comment]
> 1. I prefer use variable name for getter without a `get` prefix. `get_index ` and `get_indexes` can simply be `index` and `indexes`.

Yes, I agree, that sounds reasonable.

> 2. We don't need `error` and `code` field in response type. The helper function `deserialize_response` has already map the response to `ArangoError` if the `error` field is true.

Agree, I'll change that too.

> 3. It seems that some index type has their own fields. For instance, a geo index only has `fields` while ttl index has an additional `expire_after` field. Can we use a fat enum to ensure that a geo index must not has a `expire_after` field and etc? Also, it seems that you write several methodd to initialize `Index`. Maybe we can use a builder pattern with the help of [`typed_builder`](https://github.com/idanarye/rust-typed-builder)?

Well, that's the big question here. I did make an implementation using an `enum`. That would just mean that for every index you retrieve from the database, you were required to use a match statement before you can act on the index (call any function). If that "restriction" is Ok, we may change to an enum (internally tagged by serde).

**Comment by inzanez** [comment]
Ok, I switched to the implementation using `enum`. There's currently a `BaseIndex`, which has shared properties with `Persistent`, `Hash`, `Skiplist` and `Primary`. The others are implemented with independent structs.

I also managed to use `typed_builder` to do the `Into` conversion automatically.

**Comment by EpiFouloux** [comment]
@fMeow when can this be merged and included in a version?

**Comment by fMeow** [comment]
Sorry for the late response.

I didn't state my opinion clearly enough. For `Index`, I think we can use a struct to hold fields that are shared between all kind of index, and use enum to represent those varied across indexes.

Here is a quick but incomplete proof of concepts:
```rust
struct Index{
    // common fiedls that all kind of indexes contain 
    pub fields: Vec<String>,
    #[builder(default, setter(into))]
    pub name: Option<String>,
    #[builder(default)]
    pub id: Option<String>,
    ...
    // fields depends on index type,
    #[serde(flatten)]
    [builder(default=xxxx)]
    pub index_type: IndexType
}

enum IndexType{
   GeoIndex{geo_json:bool},
   ....
}
``` 
Though I am hesitate to name the field `index_type` and the enum `IndexType` because it is not merely a label for index type but also contains data.

Thank you!!

**Comment by fMeow** [comment]
As for the `error` and `code` field, it's exactly the field I mentioned. 

As the [HTTP guide on read index](https://www.arangodb.com/docs/stable/http/indexes-working-with.html#read-all-indexes-of-a-collection) suggests, it's used to denote whether the request succeeded. 

Let me give a quick explanation. All reponse from arangoDB can be divided into two cases: a) successful request b) failed with error code and message.  Almost all responses contain `error` and `code` even when the request succeeded, except for document related API. If a request fails, arangoDB will give back a json like `{"error":true,"code":400,"message":""}`. I think the `error` and `code` is useless for a successful request. So I use the `deserialize_response` to check whether there is a `error` field in response, and cast it to a `ClientError`.  This is done by manually deserialization of [`Response` ](https://github.com/fMeow/arangors/blob/master/src/response.rs#L47). 

Again, please feel free to remove the `error` and `code` fields. 

Thank you! :)

**Comment by inzanez** [comment]
> As for the `error` and `code` field, it's exactly the field I mentioned.
> 
> As the [HTTP guide on read index](https://www.arangodb.com/docs/stable/http/indexes-working-with.html#read-all-indexes-of-a-collection) suggests, it's used to denote whether the request succeeded.
> 
> Let me give a quick explanation. All reponse from arangoDB can be divided into two cases: a) successful request b) failed. Almost all responses contain `error` and `code` even when the request succeeded, except for document related API. If a request fails, arangoDB will give back a json like `{"error":true,"code":400,"message":""}`. I think the `error` and `code` is useless for a successful request. So I use the `deserialize_response` to check whether there is a `error` field in response, and cast it to a `ClientError`. This is done by manual deserialization of [`Response` ](https://github.com/fMeow/arangors/blob/master/src/response.rs#L47).
> 
> Again, please feel free to remove the `error` and `code` fields.
> 
> Thank you! :)

Thanks for this explanation, I removed the fields.

**Comment by inzanez** [comment]
> Sorry for the late response.
> 
> Sorry for the vague wording that I didn't state my opinion clearly enough. For `Index`, I think we can use a struct to hold fields that are shared between all kind of index, and use enum to represent those varied across indexes. The reason behind is to remove as much `Option` wrapper as possible. There is a time that almost all fields are wrapped in `Option` and I have to check if it's None for every field I read, which is terrible in my opinion.
> 
> Here is a quick but incomplete proof of concepts:
> 
> ```rust
> struct Index{
>     // common fiedls that all kind of indexes contain 
>     pub fields: Vec<String>,
>     #[builder(default, setter(into))]
>     pub name: String,
>     #[builder(default)]
>     pub id: String,
>     ...
>     // fields depends on index type,
>     #[serde(flatten)]
>     #[builder(default=xxxx)]
>     pub index_type: IndexType
> }
> 
> enum IndexType{
>    GeoIndex{geo_json:bool},
>    ....
> }
> ```
> 
> Though I am hesitate to name the field `index_type` and the enum `IndexType` because it is not merely a label for index type but also contains data.
> 
> Thank you!!

Ok, I see what you mean. But I'm not sure if it's really a more ergonomic solution in the end. Please check the following example and decide for yourself what you think might be cleaner. There's certainly less code in this one, but I think I like the abstraction concept of the current implementation better. But then that's only personal preference:

```
#[macro_use]
extern crate typed_builder;

#[macro_use]
extern crate serde;

#[derive(Debug, Serialize, Deserialize, Default, TypedBuilder)]
struct Index{
    pub fields: Vec<String>,
    #[builder(default, setter(into))]
    pub name: String,
    #[builder(default)]
    pub id: String,
    #[serde(flatten)]
    #[builder(default)]
    pub index_type: IndexType
}

#[serde(rename_all = "camelCase", tag = "type")]
#[derive(Debug, Serialize, Deserialize)]
enum IndexType{
    Persistent{unique:bool, sparse: bool, deduplicate: bool},
    Hash{unique:bool, sparse: bool, deduplicate: bool},
    Geo{geo_json:bool},
}

impl Default for IndexType {
    fn default() -> Self {
        IndexType::Persistent { unique: false, sparse: false, deduplicate: false }
    }
}

fn main() {
    let x = Index::builder().fields(vec![]).name("test").build();
    let json = serde_json::to_string(&x).unwrap();
    println!("{}", json);

    let x = Index::builder().fields(vec![]).name("test").index_type(IndexType::Geo{geo_json:true}).build();
    let json = serde_json::to_string(&x).unwrap();
    println!("{}", json);

    let json = r#"{"fields":[],"name":"test","id":"","type":"persistent","unique":false,"sparse":false,"deduplicate":false}"#;
    let x: Index = serde_json::from_str(json).unwrap();

    println!("{:#?}", x);
}
```

Please let me know what you think.

**Comment by inzanez** [comment]
I just wrote that implementation according to what you outlined above, maybe it's easier to think about it that way. I guess both would work out well.

If you prefer the current (latest) implementation, please let me finalize the documentation before merging.

**Comment by inzanez** [comment]
Documentation updated.

**Comment by fMeow** [comment]
Yes, I agree the current implement that forces to use `match` to get any data is not ergonomic. Maybe we can combine the current and previous implementation. 

```rust
#[derive(Debug, Serialize, Deserialize, Default, TypedBuilder)]
struct Index{
    pub fields: Vec<String>,
    #[builder(default, setter(into))]
    pub name: String,
    #[builder(default)]
    pub id: String,
    #[serde(flatten)]
    #[builder(default)]
    pub index_type: IndexType
}

#[serde(rename_all = "camelCase", tag = "type")]
#[derive(Debug, Serialize, Deserialize)]
enum IndexType{
    Persistent(PersistentIndex),
    Hash(HashIndex),
    Geo(GeoIndex),
}

struct PersistentIndex{unique:bool, sparse: bool, deduplicate: bool}
struct HashIndex{unique:bool, sparse: bool, deduplicate: bool}
struct GeoIndex{geo_json:bool}
impl Default for IndexType {
    fn default() -> Self {
        IndexType::Persistent(Persistent{ unique: false, sparse: false, deduplicate: false })
    }
}
```

**Comment by fMeow** [comment]
Judge from your example, it seem the Index.fileds can be empty. Then maybe we can use a default?

```rust
#[derive(Debug, Serialize, Deserialize, Default, TypedBuilder)]
struct Index{
    #[builder(default="Vec::new()")]
    pub fields: Vec<String>,
    #[builder(default, setter(into))]
    pub name: String,
    #[builder(default)]
    pub id: String,
    #[serde(flatten)]
    #[builder(default)]
    pub index_type: IndexType
}
```

**Comment by inzanez** [comment]
> Judge from your example, it seem the Index.fileds can be empty. Then maybe we can use a default?
> 
> ```rust
> #[derive(Debug, Serialize, Deserialize, Default, TypedBuilder)]
> struct Index{
>     #[builder(default="Vec::new()")]
>     pub fields: Vec<String>,
>     #[builder(default, setter(into))]
>     pub name: String,
>     #[builder(default, setter(into))]
>     pub id: String,
>     #[serde(flatten)]
>     #[builder(default)]
>     pub index_type: IndexType
> }
> ```

I just added the `default` value for the `fields`.

Regarding the implementation with `enum`: Using this I came to the conclusion that people most often use this index management to create indexes, usually on installation or something. And as this is pretty convenient, I wouldn't change it really. Looks fine I think.

**Comment by fMeow** [comment]
This PR is merged now, and will be released in the next several hours.

Thank you soooo much!

**Comment by inzanez** [comment]
Many thanks!

**Pull request merged**